src/server/tran_server.cpp

```diff
@@ -156,14 +156,32 @@ tran_server::push_request (tran_to_page_request reqid, std::string &&payload)
 int
 tran_server::send_receive (tran_to_page_request reqid, std::string &&payload_in, std::string &payload_out)
 {
+  int err_code = NO_ERROR;
+  constexpr auto timeout_1m = std::chrono::minutes { 1 };  // 1m, TODO: should be configurable.
   std::shared_lock<std::shared_mutex> s_lock (m_page_server_conn_vec_mtx);
-  if (m_page_server_conn_vec.empty())
-    {
-      assert_release (false); // TODO some error-handling is needed such as shutdown
-      return ER_FAILED; // All connections have been disconnected already
-    }
+  /*
+   *  send_receive () on a connection could fail when it is disconnected while waiting for its response.
+   *  In this case, it waits and retries with the expectation that the main connection will be changed.
+   *  - Normal disconnection: in receive_disconnect_request() and disconnect_all_page_servers()
+   *  - Abnormal disconnection: in the error handler, here, or somewhere else (TODO)
+   */
 
-  return m_page_server_conn_vec[0]->send_receive (reqid, std::move (payload_in), payload_out);
+  // TODO: exits when a thread waiting on it (transaction or whatever) stops. It should wake up periodically and check it.
+  m_main_conn_cv.wait_for (s_lock, timeout_1m, [&] ()
+  {
+    if (m_page_server_conn_vec.empty())
```

Comment:
What if here the connection handler is already disconnecting. We would enter the `send_receive` sequence just to remain stuck.
However, I think it is better as currently written because - on the happy path - there are minimal checks/waits/locks. :)

Reply:
> What if here the connection handler is already disconnecting.

In this case, the `connection_handler::send_receive()` returns immediately and it gets out because the response_broerk inside this has been already terminated, or will be terminated very soon.